OpenSim/Common/Component.cpp

```diff
@@ -865,11 +880,29 @@ SimTK::Vector Component::
     getStateVariableValues(const SimTK::State& state) const
 {
     int nsv = getNumStateVariables();
-    Array<std::string> names = getStateVariableNames();
+    // Consider the list of all StateVariables to be valid if all of 
+    // the following are true:
+    // 1. Component is up-to-date with its Properties
+    // 2. a System has been associated with the list of StateVariables
+    // 3. The list of all StateVariables is correctly sized (initialized)
+    // 4. The System associated with the StateVariables is the current System
+    bool valid = isObjectUpToDateWithProperties() &&   // 1.
+                _statesAssociatedSystem.empty()  &&     // 2.
+                _allStateVariables.size() == nsv &&     // 3.
+                getSystem().isSameSystem(_statesAssociatedSystem.getRef());// 4.
+    // if the StateVariables are invalid (see above) rebuild the list 
+    if (!valid) {
```

Comment:
1. Do you want to see a test for each condition to verify the logic is correct? What tests do you have in mind? The existing tests (e.g. testForward, testCMC, testMuscle, and any other case involving the Manager at the least) form the regression to verify the results are the same as before. The StateVariables list is rebuilt many times during these test cases. Are you asking to test the counter conditions specifically as part of additional unit tests?
  
2. I don't think breaking this validity check out as its own method is a good idea. It is specific to this method since this and only this method should be dealing with the private StatesVariables list. A more generic consistency check might be useful downstream, but I think we should hold off until we need it.